File: backend/apps/runs/progress.py
"""Whether somebody is getting better, which is the thing Pro sells.

A single round's report answers "how did that go". Only a series answers
"am I improving", and that is the question a habit product is actually
for: a report you cannot compare is a readout, and nobody renews a
readout.

Drawn for the person practising, not for a dashboard (owner's call,
2026-09-07). They want three things: proof they got better at what scared
them, one thing to work on next, and to know that a bad day is not a
relapse. So the answer carries **one row per round with the skills a
learner is trying to build** (starting at once, keeping going without
holes, holding the thread, filling the time, landing an ending, dropping a
habit word, range), and the page compares the first rounds with the last,
names the skill furthest from comfortable, and shows the floor rising. All
of it is arithmetic over what a round already stored.

Three decisions shape the lines that stay.

**Three lines and not nine.** Silence in the minute, restarts and the
opening stall are the three that improve with practice, are unambiguous,
and do not swing with the topic. Pace deliberately is not one of them: a
hot take and a story are spoken at different speeds, so a pace line drawn
across mixed genres wanders for reasons that have nothing to do with the
speaker, and a chart that says "worse" on a day somebody did fine is worse
than no chart.

**A rolling average, never the raw points.** One bad round is noise, and a
jagged line through noise reads as a verdict. The trend is smoothed over a
few rounds so a single Monday cannot make somebody think they have
regressed.

**It draws from the second round.** An earlier version waited for five and
smoothed from the first, which was wrong twice over (owner's call): the
wait put an empty box in front of somebody at exactly the moment they were
deciding whether any of this was worth having, and averaging three rounds
together when there are only three flattens the very change it is meant to
show. Two rounds is the least you can compare, and comparing two rounds is
the whole feature.

The window is the plan's, the same `streaks.Rule` the calendar and the run
list already use, so free sees five days of it and a plan sees its own
length. One rule, three features.
"""

import datetime as dt
from dataclasses import dataclass, field

from apps.runs import analysis, streaks, transcribe
from apps.runs.models import Report, Run
from apps.runs.services import owned_by

# The least that can be compared. One round is a report and not a trend;
# two is a before and an after, which is the whole point of the page.
ENOUGH = 2

# How many rounds the trend smooths over once smoothing is worth doing.
SMOOTH = 3

# Smoothing exists to make many points readable, not to hide few. Under
# this it is off entirely: averaging three rounds together when somebody
# has done four turns a real improvement into a shrug, which is the
# opposite of the job.
SMOOTH_FROM = 8

# The most points drawn. A year of practice is a thousand rounds and a line
# with a thousand points is a smear; the newest are the ones anybody is
# asking about.
MOST = 40

# Starting inside this is starting at once. The same edge the round page
# judges the stall against (`lib/report.ts`), kept in step by hand.
QUICK_START = 2.0


@dataclass(frozen=True)
class Point:
    at: dt.datetime
    stall: float
    gap: float
    fillers: float | None
    silence: float
    restarts: float


@dataclass(frozen=True)
class Minute:
    """One round drawn as the bar the done screen already draws, so the
    first and the latest can be put one above the other at the same scale.
    That side by side is the most convincing thing this product can show
    and it costs nothing beyond what is already stored."""

    at: dt.datetime
    seconds: int
    segments: list


@dataclass(frozen=True)
class Round:
    """One round as the skills a learner is building. Every field is
    arithmetic over the stored segments, transcript and word timings; a
    round nothing transcribed has the timing half and None for the rest,
    which a page tells apart from a round that scored nought."""

    id: int
    at: dt.datetime
    genre_slug: str
    prep_seconds: int
    # The speaking setting, and how much of it was used before Done or the
    # bell. Filling the time is the second against the first.
    setting: int
    spoken: int
    stall: float
    # Seconds inside the round with no voice in them: the stall and every
    # pause, and not the time after Done, which is filling the time's.
    silence: float
    gaps: int
    restarts: int
    # Whether the words carried a clock, so a nought in restarts means
    # none rather than uncounted.
    timed: bool
    ended: bool | None
    ums: int | None
    distinct: int | None
    leaned: dict[str, int] = field(default_factory=dict)
    # Whether the topic was answered, and the second the point landed.
    # Both are None on every round nothing read, which the page tells apart
    # from a round that answered nothing.
    answered: str | None = None
    point_at: float | None = None


@dataclass(frozen=True)
class First:
    at: dt.datetime
    run_id: int


@dataclass(frozen=True)
class Progress:
    enough: bool
    needed: int
    counted: int
    points: list[Point]
    first: Minute | None
    latest: Minute | None
    rounds: list[Round] = field(default_factory=list)
    firsts: dict[str, First | None] = field(default_factory=dict)


def _smooth(values: list[float | None], window: int = SMOOTH) -> list[float | None]:
    """A trailing mean over the last few readings, skipping the ones that
    are absent: a round nobody transcribed has no filler rate, and treating
    that as a zero would draw an improvement that never happened.

    A window of one is the honest identity, and that is what a short
    history gets."""
    out: list[float | None] = []
    for index in range(len(values)):
        seen = [v for v in values[max(0, index - window + 1) : index + 1] if v is not None]
        out.append(round(sum(seen) / len(seen), 2) if seen else None)
    return out


def progress(did: str, user=None, rule: streaks.Rule = streaks.FREE, *, now=None) -> Progress:
    """Every report inside the plan's window, oldest first."""
    since = (now or dt.datetime.now(dt.UTC)) - dt.timedelta(days=rule.days)
    rows = list(
        Report.objects.filter(run__in=Run.objects.filter(owned_by(did, user)), created_at__gte=since)
        .select_related("run")
        .order_by("created_at")
    )
    counted = len(rows)
    if counted < ENOUGH:
        return Progress(enough=False, needed=ENOUGH - counted, counted=counted, points=[], first=None, latest=None)

    every = [_round(row) for row in rows]
    shown = rows[-MOST:]
    recent = every[-MOST:]
    window = SMOOTH if len(shown) >= SMOOTH_FROM else 1
    # Only a provider that keeps disfluencies may contribute a filler rate.
    # A Whisper round's zero would pull the line down and read as progress
    # somebody did not make.
    rates = [row.filler_rate if row.provider == transcribe.ASSEMBLYAI else None for row in shown]
    smoothed = _smooth(rates, window)
    stalls = _smooth([row.opening_stall for row in shown], window)
    gaps = _smooth([row.longest_pause for row in shown], window)
    silences = _smooth([one.silence for one in recent], window)
    restarts = _smooth([float(one.restarts) for one in recent], window)

    points = [
        Point(
            at=row.created_at,
            stall=stalls[i] or 0.0,
            gap=gaps[i] or 0.0,
            fillers=smoothed[i],
            silence=silences[i] or 0.0,
            restarts=restarts[i] or 0.0,
        )
        for i, row in enumerate(shown)
    ]
    return Progress(
        enough=True,
        needed=0,
        counted=counted,
        points=points,
        first=_minute(rows[0]),
        latest=_minute(rows[-1]),
        rounds=every,
        firsts=_firsts(every),
    )


def _minute(row: Report) -> Minute:
    return Minute(at=row.created_at, seconds=row.run.spoken_seconds, segments=row.segments)


def _round(row: Report) -> Round:
    run = row.run
    words_at = [tuple(word) for word in row.words_at]
    has_words = row.words > 0
    honest = row.provider == transcribe.ASSEMBLYAI
    speaking = run.spoken_seconds * row.speaking_ratio
    return Round(
        id=run.pk,
        at=row.created_at,
        genre_slug=run.genre_slug,
        prep_seconds=run.prep_seconds,
        setting=run.speak_seconds,
        spoken=run.spoken_seconds,
        stall=row.opening_stall,
        silence=max(0.0, round(run.spoken_seconds - speaking, 1)),
        gaps=row.awkward_pauses,
        restarts=len(analysis.restarts(words_at)),
        timed=bool(words_at),
        ended=analysis.ended_clean(row.transcript) if has_words else None,
        ums=row.fillers if honest else None,
        distinct=analysis.distinct_words(row.transcript) if has_words else None,
        leaned=dict(analysis.leaned_on(row.transcript)) if has_words else {},
        answered=row.answered or None,
        point_at=_point_at(row, words_at),
    )


def _point_at(row: Report, words_at: list) -> float | None:
    """The second the point landed, which is the one thing about the case
    that moves with practice: the same claim made at 0:41 in August and at
    0:06 in September is the whole of what got better. None where nothing
    read the round and where a round made no point at all, never nought."""
    roles = list(row.roles)
    if "point" not in roles:
        return None
    said = analysis.sentences(row.transcript)
    spans = analysis.sentence_spans(words_at, said)
    index = roles.index("point")
    span = spans[index] if index < len(spans) else None
    return span.start if span else None


# The milestones, each the first round that met it. A round has to have
# been able to meet one for it to count: no restarts means nothing where
# no words carried a clock, and no ums means nothing where nothing counted
# them, so those two wait for a round that could have failed.
FIRSTS = {
    "no_holes": lambda r: r.gaps == 0,
    "no_restarts": lambda r: r.timed and r.restarts == 0,
    "clean_ending": lambda r: r.ended is True,
    "full_minute": lambda r: r.spoken >= r.setting,
    "quick_start": lambda r: r.stall <= QUICK_START,
    "no_ums": lambda r: r.ums == 0,
}


def _firsts(rounds: list[Round]) -> dict[str, First | None]:
    out: dict[str, First | None] = {}
    for key, met in FIRSTS.items():
        found = next((one for one in rounds if met(one)), None)
        out[key] = First(at=found.at, run_id=found.id) if found else None
    return out

File: frontend/lib/progress.ts
/* Progress, read the way somebody trying to get better reads it.

   They want three things: proof they got better at what scared them, one
   thing to work on next, and to know that a bad day is not a relapse. So
   this compares the first rounds with the last, names the skill furthest
   from comfortable, and shows the floor rising. Every function here is
   arithmetic over the rounds the backend already sends, with no DOM in it,
   so it is a table of tests rather than a page nobody can check. */

import { AWKWARD, clock, type Round } from "@/lib/report";

/** Then-and-now and the floor need this many rounds: three against three
    is the least that says anything about a person rather than a day. */
export const SIX = 6;
/** How many rounds each side of then-and-now is, at most. */
export const COMPARE_MOST = 5;
/** How many recent rounds "work on next" reads. */
export const RECENT = 5;
/** The lines draw a dot per round up to this many; past it the line
    carries it alone. */
export const DOTS_UNTIL = 12;

/* The good edges. The stall and the gap are the round page's; the
   rest are editorial, like every edge in the report, and are meant to
   move against real rounds. Silence is seconds of a round with no voice
   in them, so fifteen is a quarter of a minute quiet. */
export const GOOD_SILENCE = 15;
export const GOOD_STALL = 2;
export const GOOD_RESTARTS = 0.5;
export const GOOD_FILLED = 55;
export const GOOD_HABIT = 3;
export const GOOD_RANGE = 80;

export type SkillKey = "silence" | "stall" | "restarts" | "spoken" | "ended" | "habit" | "distinct";

export type Skill = {
  key: SkillKey;
  label: string;
  /** Higher is better. The default is lower. */
  up?: boolean;
  /** A share of rounds that managed it, not a mean. */
  rate?: boolean;
  good?: [number, number];
  scale?: [number, number];
  /** What "work on next" calls it, and what to do. A skill without these
      is never the one named: there is nothing to tell somebody to do
      about their vocabulary range. */
  name?: string;
  advice?: string;
  fmt: (value: number) => string;
  say: (then: number, now: number, of: number, word: string) => string;
};

const r1 = (value: number) => Math.round(value * 10) / 10;

export const SKILLS: Skill[] = [
  {
    key: "silence",
    label: "Silence",
    good: [0, GOOD_SILENCE],
    scale: [0, 40],
    name: "keeping going",
    advice: "Start your next point even if it is not ready. A long silence is worse than a small mistake.",
    fmt: (v) => `${Math.round(v)}s`,
    say: (a, b) => `You are quiet for ${Math.round(b)} seconds of your minute. It was ${Math.round(a)}.`,
  },
  {
    key: "stall",
    label: "Time to start",
    good: [0, GOOD_STALL],
    scale: [0, 8],
    name: "starting quickly",
    advice: "Say your point first. Then explain it.",
    fmt: (v) => `${r1(v)}s`,
    say: (a, b) => `You start after ${r1(b)} seconds. It was ${r1(a)}.`,
  },
  {
    key: "restarts",
    label: "Restarts",
    good: [0, GOOD_RESTARTS],
    scale: [0, 4],
    name: "finishing your sentences",
    advice: "If you lose a sentence, finish it anyway. Nobody heard the sentence you planned.",
    fmt: (v) => `${r1(v)} a round`,
    say: (a, b) => `You go back for a sentence ${r1(b)} times a round. It was ${r1(a)}.`,
  },
  {
    key: "spoken",
    label: "Filling the time",
    up: true,
    good: [GOOD_FILLED, 60],
    scale: [30, 60],
    name: "filling the time",
    advice: "Plan your last sentence while you think, then keep talking until the time ends.",
    fmt: (v) => clock(v),
    say: (a, b) => `You speak until ${clock(b)} of 1:00. It was ${clock(a)}.`,
  },
  {
    key: "ended",
    label: "A clean ending",
    rate: true,
    fmt: (v) => `${v}`,
    say: (a, b, of) => `You end on a full stop in ${b} of ${of} rounds. It was ${a} of ${of}.`,
  },
  {
    key: "habit",
    label: "Your word",
    good: [0, GOOD_HABIT],
    scale: [0, 12],
    name: "a habit word",
    advice: "Take a short pause where the word would go. A pause sounds calm.",
    fmt: (v) => `${r1(v)} a round`,
    say: (a, b, _of, word) => `You say "${word}" ${r1(b)} times a round. It was ${r1(a)}.`,
  },
  {
    key: "distinct",
    label: "Different words",
    up: true,
    good: [GOOD_RANGE, 140],
    scale: [40, 140],
    fmt: (v) => `${Math.round(v)}`,
    say: (a, b) => `You use ${Math.round(b)} different words in a minute. It was ${Math.round(a)}.`,
  },
];

/** A round's value on a skill, or null where the round could not be
    measured on it. A nought here always means nought. */
export function value(round: Round, skill: Skill, word: string | null): number | null {
  switch (skill.key) {
    case "silence":
      return round.silence;
    case "stall":
      return round.stall;
    case "restarts":
      return round.timed ? round.restarts : null;
    case "spoken":
      // As a share of a minute, so a two-minute setting compares with a
      // one-minute one: the skill is using the time you set.
      return round.setting > 0 ? (round.spoken / round.setting) * 60 : null;
    case "ended":
      return round.ended === null ? null : round.ended ? 1 : 0;
    case "habit":
      if (round.distinct === null || !word) return null;
      return round.leaned[word] ?? 0;
    case "distinct":
      return round.distinct;
  }
}

function measured(rounds: Round[], skill: Skill, word: string | null): number[] {
  return rounds.map((round) => value(round, skill, word)).filter((v): v is number => v !== null);
}

export function mean(values: number[]): number {
  return values.reduce((sum, v) => sum + v, 0) / values.length;
}

/** The word you lean on most across the window, which is the one worth
    tracking. Null with nothing counted. */
export function habitWord(rounds: Round[]): string | null {
  const totals: Record<string, number> = {};
  for (const round of rounds) for (const [word, count] of Object.entries(round.leaned)) totals[word] = (totals[word] ?? 0) + count;
  const ranked = Object.entries(totals).sort((a, b) => b[1] - a[1] || a[0].localeCompare(b[0]));
  return ranked.length ? ranked[0][0] : null;
}

export function inside(skill: Skill, v: number): boolean {
  return !skill.good || (v >= skill.good[0] && v <= skill.good[1]);
}

/** How far inside the comfortable stretch, 100 anywhere inside and
    falling to nothing at the end of the scale. */
export function fitOf(skill: Skill, v: number): number {
  if (!skill.good || !skill.scale || inside(skill, v)) return 100;
  const room = v > skill.good[1] ? skill.scale[1] - skill.good[1] : skill.good[0] - skill.scale[0];
  if (room <= 0) return 0;
  const over = v > skill.good[1] ? (v - skill.good[1]) / room : (skill.good[0] - v) / room;
  return Math.max(0, Math.round(100 * (1 - over)));
}

export type Compared = { skill: Skill; then: number; now: number; of: number; inside: boolean; sentence: string };

/** Your first rounds against your last, one row per skill that both sides
    could be measured on. Null under six rounds, which is the least that
    says something about a person rather than a day. */
export function thenAndNow(rounds: Round[]): { k: number; word: string | null; rows: Compared[] } | null {
  if (rounds.length < SIX) return null;
  const k = Math.min(COMPARE_MOST, Math.floor(rounds.length / 2));
  const first = rounds.slice(0, k);
  const last = rounds.slice(-k);
  const word = habitWord(rounds);
  const rows: Compared[] = [];
  for (const skill of SKILLS) {
    const a = measured(first, skill, word);
    const b = measured(last, skill, word);
    if (!a.length || !b.length) continue;
    const then = skill.rate ? a.reduce((s, v) => s + v, 0) : mean(a);
    const now = skill.rate ? b.reduce((s, v) => s + v, 0) : mean(b);
    const of = skill.rate ? Math.min(a.length, b.length) : k;
    rows.push({ skill, then, now, of, inside: inside(skill, now), sentence: skill.say(then, now, of, word ?? "") });
  }
  return { k, word, rows };
}

export type WorkOn =
  | { none: true; looked: number }
  | { none: false; skill: Skill; mean: number; out: number; looked: number; worst: Round; word: string | null };

/** The skill furthest from comfortable over the last few rounds, with the
    round that shows it worst. Only skills with something to do about
    them are candidates. */
export function workOn(rounds: Round[]): WorkOn | null {
  if (!rounds.length) return null;
  const recent = rounds.slice(-RECENT);
  const word = habitWord(rounds);
  let pick: { skill: Skill; mean: number; fit: number } | null = null;
  for (const skill of SKILLS) {
    if (!skill.advice || !skill.good) continue;
    const values = measured(recent, skill, word);
    if (!values.length) continue;
    const m = mean(values);
    const fit = fitOf(skill, m);
    if (!pick || fit < pick.fit) pick = { skill, mean: m, fit };
  }
  if (!pick || pick.fit === 100) return { none: true, looked: recent.length };
  const { skill } = pick;
  const measurable = recent.filter((round) => value(round, skill, word) !== null);
  const worst = measurable.reduce((a, b) => {
    const va = value(a, skill, word) ?? 0;
    const vb = value(b, skill, word) ?? 0;
    return (skill.up ? vb < va : vb > va) ? b : a;
  });
  const out = measurable.filter((round) => !inside(skill, value(round, skill, word) ?? 0)).length;
  return { none: false, skill, mean: pick.mean, out, looked: measurable.length, worst, word };
}

export type Floor = { worstNow: number; avgThen: number; k: number; beats: boolean };

/** Your quietest recent round against your first rounds' average. Getting
    better at this is mostly the bad days getting less bad, and this is
    true weeks before the average moves. */
export function floor(rounds: Round[]): Floor | null {
  if (rounds.length < SIX) return null;
  const k = Math.min(7, Math.floor(rounds.length / 2));
  const worstNow = Math.max(...rounds.slice(-k).map((round) => round.silence));
  const avgThen = mean(rounds.slice(0, k).map((round) => round.silence));
  return { worstNow, avgThen, k, beats: worstNow < avgThen };
}

export type GenreRow = { slug: string; rounds: number; silence: number; stall: number };

/** Rounds, silence and time to start by genre, the hardest first. Null
    with fewer than two genres, since one genre is not a comparison. */
export function byGenre(rounds: Round[]): GenreRow[] | null {
  const groups: Record<string, Round[]> = {};
  for (const round of rounds) (groups[round.genre_slug] ??= []).push(round);
  const slugs = Object.keys(groups);
  if (slugs.length < 2) return null;
  return slugs
    .map((slug) => ({
      slug,
      rounds: groups[slug].length,
      silence: mean(groups[slug].map((round) => round.silence)),
      stall: mean(groups[slug].map((round) => round.stall)),
    }))
    .sort((a, b) => b.silence - a.silence || a.slug.localeCompare(b.slug));
}

/** How you start with no time to think against how you start with some.
    Null until both have happened. */
export function byPrep(rounds: Round[]): { none: number; some: number } | null {
  const none = rounds.filter((round) => round.prep_seconds === 0);
  const some = rounds.filter((round) => round.prep_seconds > 0);
  if (!none.length || !some.length) return null;
  return { none: mean(none.map((round) => round.stall)), some: mean(some.map((round) => round.stall)) };
}

/** The milestones, in the order they are listed. */
export const FIRSTS: [string, string][] = [
  ["no_holes", "A minute with no long pauses"],
  ["no_restarts", "A round with no restarts"],
  ["clean_ending", "A clean ending"],
  ["full_minute", "A full minute"],
  ["quick_start", `Starting inside ${GOOD_STALL} seconds`],
  ["no_ums", "A minute with no ums"],
];

/** The three lines over the window, and the edge each is good under.
    Lower is better on all three, so "better" is always down. */
export const LINES = [
  { key: "silence" as const, label: "Silence in the minute", unit: "s", good: GOOD_SILENCE, floor: 40 },
  { key: "restarts" as const, label: "Restarts a round", unit: "", good: GOOD_RESTARTS, floor: 3 },
  { key: "stall" as const, label: "Time to start", unit: "s", good: GOOD_STALL, floor: 6 },
];

/** A gap long enough to notice, for the wave's warm line: the same edge
    as everywhere else. */
export const HOLE = AWKWARD;
